### PR TITLE
quic: Fix potential pointer alignmnet issue in adjustNewConnectionIdForRouting

### PR DESCRIPTION
**Pull request by RyanTheOptimist**
Casting old_connection_id_ptr to a uint32_t is not safe because there is no guarantee that old_connection_id_ptr is 32 bit aligned. Instead, just call memcpy() explicitly with the size of 4.

Risk Level: Low
Testing: Existing
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by MaskRay** [comment]
LGTM. https://maskray.me/blog/2023-01-29-all-about-undefined-behavior-sanitizer#fsanitizealignment contains links about Clang's behavior. A misaligned pointer has an unspecified value and dereferencing it is UB. Casting to `void*` `char*` with 1-byte alignment works with newer Clang.

File: source/common/quic/envoy_quic_utils.cc [hunk 1]
```diff
@@ -285,9 +285,8 @@ void adjustNewConnectionIdForRouting(quic::QuicConnectionId& new_connection_id,
                                      const quic::QuicConnectionId& old_connection_id) {
   char* new_connection_id_data = new_connection_id.mutable_data();
   const char* old_connection_id_ptr = old_connection_id.data();
-  auto* first_four_bytes = reinterpret_cast<const uint32_t*>(old_connection_id_ptr);
   // Override the first 4 bytes of the new CID to the original CID's first 4 bytes.
-  safeMemcpyUnsafeDst(new_connection_id_data, first_four_bytes);
+  memcpy(new_connection_id_data, old_connection_id_ptr, 4); // NOLINT(safe-memcpy)
```
Comment (danzh2010) [inline source/common/quic/envoy_quic_utils.cc:289]: Would replacing the reinterpret_cast with `safeMemcpyUnsafeSrc(&first_four_bytes, old_connection_id_ptr)` work?

Comment (jmarantz) [reply source/common/quic/envoy_quic_utils.cc:289]: That would only get you one byte I think.

**Review by jmarantz — APPROVED** [review]

File: source/common/quic/envoy_quic_utils.cc (hunk 1 above)
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_utils.cc:289]: Yeah, exactly. safeMemcpyUnsafeDst infers the length of the data to copy by the size of the type passed in.

**Review by danzh2010 — APPROVED** [review]

File: source/common/quic/envoy_quic_utils.cc (hunk 1 above)
Comment (RyanTheOptimist) [reply source/common/quic/envoy_quic_utils.cc:289]: Thanks, Dan! I definitely misread your first suggestion.

**Pull request merged**